src/data/helper/sourceManager.ts

```diff
@@ -346,8 +350,79 @@ export class SourceManager {
      * @param sourceIndex By defualt 0, means "main source".
      *                    Most cases there is only one source.
      */
-    getSource(sourceIndex?: number) {
-        return this._sourceList[sourceIndex || 0];
+    getSource(sourceIndex?: number): Source {
+        sourceIndex = sourceIndex || 0;
+        const source = this._sourceList[sourceIndex];
+        if (!source) {
+            // Series may share source instance with dataset.
+            const upSourceMgrList = this._getUpstreamSourceManagers();
```

Comment:
This optimization is based on the profile result. In fact, in the case of #11907, The most memory cost thing is the storage of dimensions info. Previously we need to store 1000 dimensions in each series, which is 1 million in total. It's not only in the `List` instance. But also in the `Source` instance.

`createSource` will also parse all dimensions and store them in the `Source` instance. If we don't share them, the heap memory will increase from 20MB to 140MB and the render time will also be much longer. What makes it worse is it increases exponentially.

Sharing source may make things bit more complex. But I think it won't break the design of the dataset. And the good thing is it can be all done in the SourceManager and won't leak the abstraction to other modules.

Reply:
Oh, get it.
Do we need to add some brief comment or the link of this explantation in source code? Other one may not get the reason of this consideration when reading code.